3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,222 @@
+package common.enrichments.registry
+
+// Scala
+import org.json4s.JValue
+import scala.collection.JavaConverters._
+
+// Snowplow
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.jayway.jsonpath.MapFunction
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.ValidatedNelMessage
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+
+// Scala libraries
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+
+// Java
+import java.security.{MessageDigest, NoSuchAlgorithmException}
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+import com.github.fge.jsonschema.core.report.ProcessingMessage
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath, Option => JOption}
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion
+
+// Scalaz
+import scalaz.Scalaz._
+import scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ */
+sealed trait PiiField {
+  val strategy: PiiStrategy
+  val fieldName: String
+  def applyStrategy(fieldValue: String): String
+  val strategyApplication =
+    (v: Validation[String, String]) => v.flatMap(applyStrategy(_).success)
+  def transformer(transformMap: TransformMap) =
+    transformMap.collect {
+      case (inputField: String, (tf: Function2[String, String, Validation[String, String]], outputField: String))
+          if (outputField == fieldName) =>
+        (inputField, ((arg1: String, arg2: String) => tf.tupled.andThen(strategyApplication)((arg1, arg2)), outputField))
+    }
+}
+
+/**
+ * PiiStrategy trait. This corresponds to a strategy to apply to a single field. Currently only only String input is
+ * supported.
+ */
+trait PiiStrategy {
+  val scramble: String => String
+}
+
+/**
+ * Companion object. Lets us create a PiiPseudonymizerEnrichment
+ * from a JValue.
+ */
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+  val supportedSchema = SchemaCriterion("com.snowplowanalytics.snowplow.enrichments", "pii_enrichment_config", "jsonscehma", 1, 0, 0)
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    for {
+      conf      <- matchesSchema(config, schemaKey)
+      piiFields <- ScalazJson4sUtils.extract[List[JObject]](conf, "parameters", "pii").leftMap(_.getMessage)
+      strategyFunction <- ScalazJson4sUtils
+        .extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction")
+        .leftMap(_.getMessage)
+      hashFunction <- try {
```

Comment:
Can we extract all these 3-4-liners into separate private functions? It will reduce `parse` size to 6-8 lines.